### PR TITLE
WIP: Register derivedFrom and svd updates

### PR DESCRIPTION
**Pull request by jamesmunns**
This PR extends https://github.com/rust-embedded/svd2rust/pull/193 (edit: and https://github.com/rust-embedded/svd2rust/pull/190), and also brings support for the changes made in `svd-parser`, primarily understanding the `RegisterCluster` type.

### CODE REVIEW

**Comment by jamesmunns** [comment]
bors try

**Comment by jamesmunns** [comment]
Note, I believe this failure may be a regression of `svd`'s ability to parse MSP430 SVDs. The backtrace for this failure (run locally) looks like this:

```text
➜  other-test git:(register-derived-from) ✗ RUST_BACKTRACE=full ../target/release/svd2rust --target msp430 -i ./msp430g2553.svd
thread 'main' panicked at 'called `Option::unwrap()` on a `None` value', libcore/option.rs:355:21
stack backtrace:
   0:     0x55dbd43a19b3 - std::sys::unix::backtrace::tracing::imp::unwind_backtrace::h76492c069e044094
                               at libstd/sys/unix/backtrace/tracing/gcc_s.rs:49
   1:     0x55dbd439c488 - std::sys_common::backtrace::_print::h94a018b30274f0aa
                               at libstd/sys_common/backtrace.rs:71
   2:     0x55dbd43a0b74 - std::panicking::default_hook::{{closure}}::ha3ab490579270e44
                               at libstd/sys_common/backtrace.rs:59
                               at libstd/panicking.rs:211
   3:     0x55dbd43a08dd - std::panicking::default_hook::hc3ba68bed94f8a90
                               at libstd/panicking.rs:227
   4:     0x55dbd43a11fe - std::panicking::rust_panic_with_hook::h8380fbfdb3a9ef26
                               at libstd/panicking.rs:476
   5:     0x55dbd43a0da1 - std::panicking::continue_panic_fmt::he5272a22537494ee
                               at libstd/panicking.rs:390
   6:     0x55dbd43a0c85 - rust_begin_unwind
                               at libstd/panicking.rs:325
   7:     0x55dbd43b1d0c - core::panicking::panic_fmt::he3157e9749b52890
                               at libcore/panicking.rs:77
   8:     0x55dbd43b1c3b - core::panicking::panic::h629a30fc6a0298aa
                               at libcore/panicking.rs:52
   9:     0x55dbd4305632 - <svd_parser::svd::device::Device as svd_parser::parse::Parse>::parse::h98591c7a661028b6
  10:     0x55dbd42f0a01 - svd_parser::parse::h534a94050a300c22
  11:     0x55dbd42202a7 - svd2rust::main::h34a3bf77a67818e0
  12:     0x55dbd422f3d2 - std::rt::lang_start::{{closure}}::hc6a90ef6c35166bc
  13:     0x55dbd43a0c22 - std::panicking::try::do_call::h6d19b60204e9d839
                               at libstd/rt.rs:59
                               at libstd/panicking.rs:310
  14:     0x55dbd43aeac9 - __rust_maybe_catch_panic
                               at libpanic_unwind/lib.rs:102
  15:     0x55dbd43a1533 - std::rt::lang_start_internal::h62a3cf6eecce6831
                               at libstd/panicking.rs:289
                               at libstd/panic.rs:398
                               at libstd/rt.rs:58
  16:     0x55dbd4224cc4 - main
  17:     0x7efcaa486222 - __libc_start_main
  18:     0x55dbd421319d - _start
  19:                0x0 - <unknown>
```
CC @ryankurte @Emilgardis, I'll dig in to this a bit, and if I find anything, I'll report it upstream to `svd`.

**Comment by jamesmunns** [comment]
I'd like @wez to review this before merging, to make sure this change is still necessary. There's some confusion described in https://github.com/nrf-rs/nrf52832-pac/pull/11#issuecomment-439653552

**Comment by Emilgardis** [comment]
ping @wez

**Comment by wez** [comment]
Sorry; have been swamped over the Thanksgiving holidays.  I'll take a look over the weekend

**Comment by wez** [comment]
I tested this with https://github.com/atsamd-rs/atsamd/blob/master/update.sh and am sorry to report that it doesn't appear to correctly generate the pmux array.  If you run that script and then `git diff` the repo, you'll see that the `dim` related attributes don't seem to be inherited,  so rather than generating an array, it generates a sequence of wrongly sized registers that it then skips because the size sanity check doesn't pass.

It's not clear to me what the cause of this is; it looks like you pulled in the commits that made it work as-is, and then made some seemingly unrelated changes, so I suspect that something else changed in svd2rust or svd that needs to be accounted for.  I didn't have time to fully dig into this

**Comment by wez** [comment]
Ok, I traced the issue; the SVD file declares derivedFrom as an attribute:

```
        <register derivedFrom="PMUX0_%s">
```

In my original commit to svd: https://github.com/wez/svd/commit/211ae3e42b78a0ab851ffc9a23b343ce7b9bed48 this was being extracted correctly, but in the current svd code this appears to have been accidentally refactored to look at a child element instead.

This patch to the svd parser is needed to fix things:

```
$ git diff
diff --git a/src/svd/registerinfo.rs b/src/svd/registerinfo.rs
index a075e38..6cab32f 100644
--- a/src/svd/registerinfo.rs
+++ b/src/svd/registerinfo.rs
@@ -59,7 +59,7 @@ impl RegisterInfo {
             name,
             alternate_group: tree.get_child_text_opt("alternateGroup")?,
             alternate_register: tree.get_child_text_opt("alternateRegister")?,
-            derived_from: tree.get_child_text_opt("derivedFrom")?,
+            derived_from: tree.attributes.get("derivedFrom").map(|s| s.to_owned()),
             description: tree.get_child_text("description")?,
             address_offset: tree.get_child_u32("addressOffset")?,
             size: parse::optional::<u32>("size", tree)?,
```

**Comment by wez** [comment]
I'd love to see this land, because I spent an hour basically re-creating this PR tonight, thinking that it shipped already and that something else was broken with svd2rust#master and svd#master :-/

I think the following need to happen to get this landed:

* svd-parser needs to release 0.7 with https://github.com/japaric/svd/pull/65 included (0.7 because of the RegisterCluster and other api changes)
* this PR needs to point to that

Also, I wanted to note that these `derivedFrom` changes are closing in on 1 year since they were submitted and are still waiting for a release.  What help do you need to get this done?

File: Cargo.toml [hunk 1]
```diff
@@ -33,5 +33,5 @@ either = "1.0.3"
 error-chain = "0.11.0"
 inflections = "1.1.0"
 quote = "0.3.15"
-svd-parser = "0.6"
+svd-parser = { git = "https://github.com/japaric/svd", rev = "6aa4fdc138b91d7b9d2f0bd42fd1c85cb4fb5df9" }
```
Comment (ryankurte) [inline Cargo.toml]:
```suggestion
svd-parser = "0.7.0"
```

**Comment by ryankurte** [comment]
hey @jamesmunns are you able to look at these merge conflicts?

**Comment by Emilgardis** [comment]
bors try

**Comment by ryankurte** [comment]
hmm [bors queue](https://app.bors.tech/repositories/2917) says pending, but there's a [crash](https://app.bors.tech/repositories/2917/log) in the log. i found a [closed issue](https://github.com/bors-ng/bors-ng/issues/398) about not working with branch protection on the staging branch, but, i'm not sure why that would be an issue in this instance.

bors try

**Comment by Emilgardis** [comment]
bors try-

bors try

**Comment by therealprof** [comment]
@jamesmunns Mind rebasing this?

**Comment by burrbull** [comment]
Conflicts? Again?

**Comment by therealprof** [comment]
@burrbull Yeah, just got two PRs merged, that's the one in `src/generate/register.rs`. `src/main.rs` is the one which removed `extern crate` which I though you might have caught.

**Comment by burrbull** [comment]
Pull from https://github.com/burrbull/svd2rust/tree/register-derived-from

**Comment by burrbull** [comment]
#313

**Comment by burrbull** [comment]
cc @jamesmunns 
Please, rebase this PR, or approve #314 .

**Comment by adamgreig** [comment]
Closing in preference of #319 which incorporates this PR.

**Pull request closed without merging**